doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -848,37 +1231,425 @@ The Ltac2 parser can be extended with syntactic notations.
+  which is not accepted by Ltac2.  This is because Ltac2 uses a different
+  definition for :token:`delta_flag`; compare it to :token:`ltac2_delta_flag`.  This also affects
+  :tacn:`compute`.
+
+Ltac1 tactics are not automatically available in Ltac2.  (Note that some of the tactics described
+in the documentation are defined with Ltac1.)
+You can make them accessible in Ltac2 with the following commands:
+
+.. coqtop:: in
+
+   From Coq Require Import Lia.
+   Local Ltac2 lia_ltac1 () := ltac1:(lia).
+   Ltac2 Notation "lia" := lia_ltac1 ().
+
+A similar approach can be used to access missing built-in tactics.  See :ref:`simple_api` for an
+example that passes two parameters to a missing build-in tactic.
+
+.. _syntactic_classes:
+
+Syntactic classes
+~~~~~~~~~~~~~~~~~
+
+The simplest syntactic classes in Ltac2 notations represent individual nonterminals
+from the |Coq| grammar.  Only a few selected nonterminals are available as syntactic classes.
+In addition, there are metasyntactic operations for describing
+more complex syntax, such as making an item optional or representing a list of items.
+When parsing, each syntactic class expression returns a value that's bound to a name in the
+notation definition.
+
+Syntactic classes are described with a form of S-expression:
+
+   .. insertprodn ltac2_scope ltac2_scope
+
+   .. prodn::
+      ltac2_scope ::= @string
+      | @int
+      | @name
+      | @name ( {+, @ltac2_scope } )
+
+.. todo no syn class for ints or strings?
+  parm names are not reserved (e.g the var can be named "list1"
+
+Metasyntactic operations that can be applied to other syntactic classes are:
+
+  :n:`opt(@ltac2_scope)`
+    Parses an optional :token:`ltac2_scope`.  The associated value is either :n:`None` or
+    enclosed in :n:`Some`
+
+  :n:`list1(@ltac2_scope {? , @string })`
+    Parses a list of one or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.
+
+  :n:`list0(@ltac2_scope {? , @string })`
+    Parses a list of zero or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
+    items must be separated by :token:`string`.  For zero items, the associated value
+    is an empty list.
+
+  :n:`seq({+, @ltac2_scope })`
+    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
+    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
+    `self` and `next` are not permitted within `seq`.
+
+The following classes represent nonterminals with some special handling.  The
+table further down lists the classes that that are handled plainly.
+
+  :n:`constr {? ( @scope_key ) }`
+    Parses a :token:`term`.  If specified, the :token:`scope_key` is used to interpret
+    the term (as described in  :ref:`LocalInterpretationRulesForNotations`)
+
+    .. todo PR the original text describes "constr:(c%A%B)" which I couldn't make sense of
+       Is that something we should keep (and if so, what is an example?)
```

Comment:
Yes, we should keep it.  An example is
```coq
Require Import Ltac2.Ltac2.

Ltac2 Notation "foo" c(constr(nat)) := Message.print (Message.of_constr c).
Ltac2 Notation "bar" c(constr(type)) := Message.print (Message.of_constr c).

Goal True.
  foo (1+1).
  Fail bar (1+1). (* The term "1" has type "nat" while it is expected to have type "Type". *)
  bar (nat+nat). (* (nat + nat)%type *)
  Fail foo (nat+nat). (* The term "nat" has type "Set" while it is expected to have type "nat". *)
```
An example of nested scopes is
```coq
Require Import Ltac2.Ltac2.
Require Import Coq.Strings.String.
Require Import Coq.Lists.List.

Ltac2 Notation "foo" c(constr(string,list)) := Message.print (Message.of_constr c).

Goal True.
  foo "a".
  foo ("a" ++ "a").
  foo (1 :: nil).
  Fail foo (@nil nat ++ nil).
  foo (@nil nat ++ nil)%list.
```
where doing `constr(A,B)` is equivalent to interpreting in the scope stack `%A%B`.  (Note that your syntax is wrong, it allows a comma-delimited list.)

Reply:
> (Note that your syntax is wrong, it allows a comma-delimited list.)

I corrected this, but I don't think an example is necessary.  Not easy to get the syntax since this isn't in the mlg files but only in OCaml code that I don't fully understand:

```
let () = add_scope "constr" (fun arg ->
    let delimiters = List.map (function
        | SexprRec (_, { v = Some s }, []) -> s
        | _ -> scope_fail "constr" arg)
        arg
    in
    let act e = Tac2quote.of_constr ~delimiters e in
    Tac2entries.ScopeRule (Pcoq.Symbol.nterm Pcoq.Constr.constr, act)
  )
```